modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java

```diff
@@ -176,26 +179,187 @@
 import org.elasticsearch.painless.symbol.Decorations.UnaryType;
 import org.elasticsearch.painless.symbol.Decorations.UpcastPainlessCast;
 import org.elasticsearch.painless.symbol.Decorations.ValueType;
+import org.elasticsearch.painless.symbol.FunctionTable;
 import org.elasticsearch.painless.symbol.FunctionTable.LocalFunction;
 import org.elasticsearch.painless.symbol.ScriptScope;
 import org.elasticsearch.painless.symbol.SemanticScope.Variable;
+import org.objectweb.asm.Opcodes;
 
+import java.lang.invoke.CallSite;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.MethodType;
 import java.lang.reflect.Modifier;
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
 import java.util.regex.Pattern;
 
-public class DefaultUserTreeToIRTreeVisitor implements UserTreeVisitor<ScriptScope> {
+public class DefaultUserTreeToIRTreePhase implements UserTreeVisitor<ScriptScope> {
 
-    private ClassNode irClassNode;
+    protected ClassNode irClassNode;
 
-    protected IRNode visit(ANode userNode, ScriptScope scriptScope) {
-        if (userNode == null) {
-            return null;
-        } else {
-            userNode.visit(this, scriptScope);
-            return scriptScope.getDecoration(userNode, IRNodeDecoration.class).getIRNode();
+    /**
+     * This injects additional ir nodes required for resolving the def type at runtime.
+     * This includes injection of ir nodes to add a function to call
+     * {@link DefBootstrap#bootstrap(PainlessLookup, FunctionTable, Lookup, String, MethodType, int, int, Object...)}
+     * to do the runtime resolution.
+     */
+    protected void injectBootstrapMethod(ScriptScope scriptScope) {
```

Comment:
Should you mention `injectStaticFields` here?

Reply:
Done.